### ISSUE TITLE
Improve docstring for test_notTrapped in twisted.test.test_failure

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#5952]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)| @itamarst reported|
|-|-|
|Trac ID|trac#5952|
|Type|enhancement|
|Created|2012-09-07 12:21:34Z|
|Branch|https://github.com/twisted/twisted/tree/test-not-trapped-5952|

The docstring for `test_notTrapped` doesn't actually describe what the test does very well; it should be rewritten to explain that.

Attachments:

* [my-first-patch.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/ee0/ee07861ec1b5b66ff87ab5adb0c412ef9a6e549f/3eb8670a2085c41df03fbf839047280cd1c6f46b.patch) (1907 bytes) - added by sreepriya on 2013-03-04 08:55:45Z - In the patch, the method test_notTrapped() is broken to two methods for PY3 and PY2. Docstrings are added for both the methods.  
* [patch1.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/ee0/ee07861ec1b5b66ff87ab5adb0c412ef9a6e549f/881ec6bc8ac29284f30b2343ee14193d0a56ffba.patch) (1765 bytes) - added by sreepriya on 2013-03-05 14:18:15Z - This patch breaks the method names test_notTrapped into two methods and docstring added for both. The Comment11  review changes are implemented. 
* [patch3.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/ee0/ee07861ec1b5b66ff87ab5adb0c412ef9a6e549f/674de2138d6d54a3df372b7a9df60eea7a429e25.patch) (2711 bytes) - added by sreepriya on 2013-03-23 06:03:49Z - The patch includes the comments given by rwall. Coding standards are followed, old test_trapped is removed and broken to two functions, if_PY3 is made closer. 

<details><summary>Searchable metadata</summary>

```
trac-id__5952 5952
type__enhancement enhancement
reporter__itamar itamar
priority__normal normal
milestone__ 
branch__branches_test_not_trapped_5952 branches/test-not-trapped-5952
branch_author__rwall rwall
status__closed closed
resolution__fixed fixed
component__core core
keywords__easy_documentation easy documentation
time__1347020494000000 1347020494000000
changetime__1365795809000000 1365795809000000
version__None None
owner__rwall rwall
cc__thijs cc__sreepriya
```
</details>

**Comment by twisted-trac** [comment]
|<img alt="sreepriya's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|sreepriya set owner to sreepriya|
|-|-|

**Comment by twisted-trac** [comment]
|<img alt="sreepriya's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_4"></a>|sreepriya commented|
|-|-|

Instead of the docstring "Making sure trap doesn't trap what it shouldn't", does the following docstring make sense ? 
"Avoids trapping failures that are defined"

**Comment by twisted-trac** [comment]
|[<img alt="tomprince's avatar" src="https://avatars.githubusercontent.com/u/283816?s=50" width="50" height="50">](https://github.com/tomprince)<a name="note_6"></a>|@tomprince commented|
|-|-|

I'm not sure if that is much better. Unfortunately, there isn't much documentation on what makes a good test docstring (see [[#6301](https://github.com/twisted/twisted/issues/6301)](https://github.com/twisted/twisted/issues/6301), which links to [#6301 this google+ post]). 

Generally a test docstring wants to be something like `When <some thing is done>, <something else occurs>`. Often `<something is done>` is a method being called. And `<something else occurs>` is generally what is captured by the assertion(s).

In this particular case, I think the method may actually want to be split up into two methods. One for the python 2 only assertion, and one for the python 2+3 assertion.

**Comment by twisted-trac** [comment]
|<img alt="sreepriya's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_7"></a>|sreepriya commented|
|-|-|

If a new method is added, I must also make changes in the function names in the places where these tests are conducted right?

**Comment by twisted-trac** [comment]
|<img alt="sreepriya's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_8"></a>|sreepriya commented|
|-|-|

Why do we need a separate assertion for python 2 ? Why is it that python 2+3 doesn't work for even python 2 ?

**Comment by twisted-trac** [comment]
|<img alt="sreepriya's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|sreepriya set owner to @tomprince|
|-|-|

**Comment by twisted-trac** [comment]
|<img alt="sreepriya's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|sreepriya removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="tomprince's avatar" src="https://avatars.githubusercontent.com/u/283816?s=50" width="50" height="50">](https://github.com/tomprince)<a name="note_11"></a>|@tomprince set owner to sreepriya|
|-|-|

Thanks for your contribution.
  1. There are a number of unrelated changes in this patch, and it doesn't apply cleanly to trunk. Also `test_notTrapped_PY3` is weirdly indented.
  2. The method being tested here is `Failure.trap`, so the docstring should say something about the expected behaviour of that function. You should probably read the [api documentation](https://twistedmatrix.com/documents/current/api/twisted.python.failure.Failure.html#trap) for that method, as well as the relevant [howto](https://twistedmatrix.com/documents/current/core/howto/defer.html#auto4). Also, the comments in the method are probably helpful.
  3. The method names don't follow the [coding standard](https://twistedmatrix.com/documents/current/core/development/policy/coding-standard.html#auto17). Also, they test something slight different, and should be named accordingly.
  4. Rather than doing `if _PY3` in the test, it should be done outside, setting `.skip` on the test, with a message describing why it is being skipped.
  5. (minor) There should be 2 blank lines between methods in a class.
Please resolve the above issues and resubmit for review.

**Comment by twisted-trac** [comment]
|<img alt="sreepriya's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|sreepriya removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="wallrj's avatar" src="https://avatars.githubusercontent.com/u/978965?s=50" width="50" height="50">](https://github.com/wallrj)<a name="note_14"></a>|@wallrj set owner to @wallrj<br>@wallrj set status to `assigned`|
|-|-|

Reviewing...

**Comment by twisted-trac** [comment]
|[<img alt="wallrj's avatar" src="https://avatars.githubusercontent.com/u/978965?s=50" width="50" height="50">](https://github.com/wallrj)<a name="note_15"></a>|@wallrj commented|
|-|-|

(In [37643]) Branching to 'test-not-trapped-5952'

**Comment by twisted-trac** [comment]
|[<img alt="wallrj's avatar" src="https://avatars.githubusercontent.com/u/978965?s=50" width="50" height="50">](https://github.com/wallrj)<a name="note_16"></a>|@wallrj removed owner<br>@wallrj set status to `new`|
|-|-|

Code Review:

Thanks sreepriya for your latest patch.

I created a new branch for you and applied your patch. See
 * log:branches/test-not-trapped-5952

Here are some comments:

 1. [Build Results seem ok](http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/test-not-trapped-5952)
 2. There is trailing whitespace in your patch - please remove it.
 3. The lines in the new docstrings are too long. Please rewrap to <
    80 columns.
 4. The news file was missing. I added one for you in the branch.
 5. The new test method names don't follow the coding standard. The
    first letter after "test_" should be lower case eg
    test_trappedAndReRaiseFailure. See existing test methods for
    examples.
 6. I don't think you addressed Tom Prince comments about the test
    docstrings. They are still not clear to me. Test docstrings should
    explain the expected behaviour of failure.trap. They currently
    seem to describe how the test works.
 7. The patch replaces the old test_notTrapped with two new tests but
    you haven't removed the old test.
 8. I think you should move the if not __PY3 skip closer to the skipped
    test.
 9. Also add a more specific reason for skipping the test with PY3

Please resubmit your patch against the new branch.

Thanks again.

**Comment by twisted-trac** [comment]
|[<img alt="wallrj's avatar" src="https://avatars.githubusercontent.com/u/978965?s=50" width="50" height="50">](https://github.com/wallrj)<a name="note_17"></a>|@wallrj commented|
|-|-|

Here's exarkun's explanation for the PY3 test skip:

```
17:46 < exarkun> Failure.trap is stu0000000000000pid
17:47 < exarkun> Yes, you cannot raise Failure on Python 3, so it doesn't.  And you should never have raised Failure on Python 2 anyway, so it's too 
                 bad that Failure does.
```

```
>>> class Foo:
...     def __init__(self, *args):
...             pass
... 
>>> raise Foo('test')
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: exceptions must derive from BaseException

```

So instead of "This test works only with Python 2" say something like "In Python3, failure.trap raises the original Exception instead of a failure instance because Python3 can only raise BaseException subclasses."

**Comment by twisted-trac** [comment]
|[<img alt="wallrj's avatar" src="https://avatars.githubusercontent.com/u/978965?s=50" width="50" height="50">](https://github.com/wallrj)<a name="note_19"></a>|@wallrj commented|
|-|-|

(In [37806]) apply patch3.patch from sreepriya. refs [#5952](https://github.com/twisted/twisted/issues/5952)

**Comment by twisted-trac** [comment]
|[<img alt="wallrj's avatar" src="https://avatars.githubusercontent.com/u/978965?s=50" width="50" height="50">](https://github.com/wallrj)<a name="note_20"></a>|@wallrj commented|
|-|-|

(In [37808]) rewrap the skip message. refs [#5952](https://github.com/twisted/twisted/issues/5952)

**Comment by twisted-trac** [comment]
|[<img alt="wallrj's avatar" src="https://avatars.githubusercontent.com/u/978965?s=50" width="50" height="50">](https://github.com/wallrj)<a name="note_21"></a>|@wallrj commented|
|-|-|

(In [37811]) the python3 skip test was referencing the wrong function and was in the wrong place. refs [#5952](https://github.com/twisted/twisted/issues/5952)

**Comment by twisted-trac** [comment]
|[<img alt="wallrj's avatar" src="https://avatars.githubusercontent.com/u/978965?s=50" width="50" height="50">](https://github.com/wallrj)<a name="note_22"></a>|@wallrj commented|
|-|-|

(In [37812]) a better test method name and a clearer docstring. refs [#5952](https://github.com/twisted/twisted/issues/5952)

**Comment by twisted-trac** [comment]
|[<img alt="wallrj's avatar" src="https://avatars.githubusercontent.com/u/978965?s=50" width="50" height="50">](https://github.com/wallrj)<a name="note_23"></a>|@wallrj commented|
|-|-|

(In [37813]) a better test method name and a clearer docstring. refs [#5952](https://github.com/twisted/twisted/issues/5952)

**Comment by twisted-trac** [comment]
|[<img alt="wallrj's avatar" src="https://avatars.githubusercontent.com/u/978965?s=50" width="50" height="50">](https://github.com/wallrj)<a name="note_24"></a>|@wallrj commented|
|-|-|

Thanks sreepriya

I reviewed your latest [patch3.patch](None/patch3.patch) and fixed some things
at the same time.

 * There was trailing whitespace and an unrelated deletion of a blank
   line - Fixed.
 * The python3 skip test was referencing the old test method name and
   using the class name within the class definition - Fixed.
 * I think you misunderstood my previous review comment about the test
   method docstrings. I've rewritten the docstrings and given the test
   methods clearer names (IMHO).
 * I simplified the test for trap raising the wrapped Exception and made
   it a Python3 only test.

Ready for review in log:branches/test-not-trapped-5952
 * [Build results](http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/test-not-trapped-5952)

**Comment by twisted-trac** [comment]
|[<img alt="tomprince's avatar" src="https://avatars.githubusercontent.com/u/283816?s=50" width="50" height="50">](https://github.com/tomprince)<a name="note_25"></a>|@tomprince set owner to @wallrj|
|-|-|

1.You can probably just say "subclass of one of the expected", rather than having a parenthetical.
 2. The test incidentally checked that `failure.Failure()`, when catching a `Failure` actually creates a `Failure` with the right `.value` (i.e. the exception not the caught `Failure`). It would be good to add a test to do that. Although, perhaps it would be better to simply construct a failure with another. Or perhaps 2, a) construct a failure with a failure, b) `.trap` -> `Failure()`.

**Comment by twisted-trac** [comment]
|[<img alt="wallrj's avatar" src="https://avatars.githubusercontent.com/u/978965?s=50" width="50" height="50">](https://github.com/wallrj)<a name="note_26"></a>|@wallrj removed owner|
|-|-|

Thanks for the review Tom.  

I've responded to your comments below...

Replying to [tom.prince](#note_25):

>  1. You can probably just say "subclass of one of the expected",
>  rather than having a parenthetical.

Done.

>  2. The test incidentally checked that `failure.Failure()`, when
>  catching a `Failure` actually creates a `Failure` with the right
>  `.value` (i.e. the exception not the caught `Failure`). It would be
>  good to add a test to do that. Although, perhaps it would be better
>  to simply construct a failure with another. Or perhaps 2, a)
>  construct a failure with a failure, b) `.trap` -> `Failure()`.

I added two new tests as suggested. But actually I think that the
second case is already handled in FindFailureTests.

Ready for another review in log:branches/test-not-trapped-5952

 * [Build Results](http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/test-not-trapped-5952)

**Comment by twisted-trac** [comment]
|[<img alt="tomprince's avatar" src="https://avatars.githubusercontent.com/u/283816?s=50" width="50" height="50">](https://github.com/tomprince)<a name="note_27"></a>|@tomprince set owner to @wallrj|
|-|-|

Looks good, please merge.

**Comment by twisted-trac** [comment]
|[<img alt="wallrj's avatar" src="https://avatars.githubusercontent.com/u/978965?s=50" width="50" height="50">](https://github.com/wallrj)<a name="note_28"></a>|@wallrj set status to `closed`|
|-|-|

(In [38052]) Merge test-not-trapped-5952: Improved tests and test documentation for Failure.trap.

Author: sreepriya, rwall
Reviewer: tom.prince, rwall
Fixes: [#5952](https://github.com/twisted/twisted/issues/5952)

Improved tests and test documentation for Failure.trap.
